Added types and improved CS

nomack84 opened the pull request:
I've added types to several types and fixed some CS issues.

bryanrsmith commented on src/router.js:
```diff
@@ -80,15 +80,14 @@ export class Router {
   */
-  registerViewPort(viewPort:Object, name?:string) {
-    name = name || 'default';
+  registerViewPort(viewPort : Object, name? : string = 'default') : void {
```
I'm a little nervous about this change. I think templating-router may end up passing empty string to this, and this change would break that. Have you checked out existing callers of this method?

nomack84 replied:
I didn't think about the possibility of empty strings, so this indeed will break the code. I'll revert this change.

nomack84 replied:
It doesn't break it, because `default` is assumed only when `undefined` is passed. You can check it out. 

```javascript
function defaults(x = 'default') {
  console.log(x);
}

defaults(''); -> prints ''
defaults('view'); -> prints 'view'
defaults(); -> prints 'default'
```
Check here: http://www.es6fiddle.net/idjr4tkv/

nomack84 replied:
What could be a valid argument against default params is the code Babel generates.
```javascript 
function defaults(x = 'default') {
  console.log(x);
}
```
generates:
```javascript
function defaults() {
  var x = arguments.length <= 0 || arguments[0] === undefined ? 'default' : arguments[0];

  console.log(x);
}
```
Looking at this, perhaps we is better to use the 'old' way. What do you think?

bryanrsmith replied:
I don't mind the generated code, I just want to make sure that the code in aurelia/templating-router never passes null or empty string and expects the default to be applied. I am not sure what how behavior will change in the cases of `<router-view></router-view>` and `<router-view name=""></router-view>`.

nomack84 replied:
Well, I checked the behavior in the above fiddle and it doesn't fail. If an empty string is passed, it doesn't use the `default`. So I don't see why it'll fail. Is there any test that cover this, or perhaps I should add it...

bryanrsmith replied:
Right... viewPorts shouldn't be allowed to have an empty string for a name. If empty string is passed, it should get the default.

bryanrsmith commented:
Thanks, @nomack84. Would you mind updating your commit message to match our contributing guidelines?

These changes look great. If you're interested in helping us improve code quality, we're just starting to roll out more comprehensive style and linter config. You can check out a PR that makes that change in the metadata repo here: aurelia/metadata#20 (The PR is out of date now, but the config changes and such are still valid).

nomack84 commented:
@bryanrsmith I love a lot Aurelia and admire a lot the great work you guys are doing. Even with such a great difference in resources and people, you are doing much better than Angular, at least that's what I think. So definitely if I can do at least something as simple as improving code quality, certainly can count on me.

nomack84 commented:
By the way, where can I find the contributing guidelines?

nomack84 commented:
Forgot it, I've found them :)

bryanrsmith commented:
Thank you for the kind words. We would of course appreciate any contributions you might have time for.

https://github.com/DurandalProject/about/blob/master/CONTRIBUTING.md#commit

EisenbergEffect commented:
CLA confirmed. @bryanrsmith Proceed as you deem appropriate.

bryanrsmith commented:
@nomack84 I think we should revert the default value change. Other than that this looks good to me. Could you please rebase against master, squash your commits, and update the commit message?

nomack84 commented:
@bryanrsmith I've reverted the change and updated the commit message. Can you help me rebasing against master? Any guidance about how to achive it?
Thanks!

bryanrsmith commented:
Sure. Here's a good explanation with steps to take: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

Let me know if I can help further.

nomack84 commented:
@bryanrsmith I've rebased against master already.

bryanrsmith commented:
It looks like it's still out of date. I was able to resolve the conflicts myself, but unfortunately the circular dependencies added in this PR (for type annotations) is preventing it from building (`gulp build`). We use a tool that calculates concatenation order based on import statements, and it doesn't handle circular dependencies at the moment. I'll try to take a look at that this weekend. I won't be able to merge this until I've got that fixed.

bryanrsmith commented:
@nomack84 Nevermind. I was able to get it working by removing the imports that were only needed for type annotations. I went ahead and resolved the conflicts myself and pushed the change to master -- 8e528c858c8aacc5a5f8c7f087516de1e160d5d9

Thanks for the contribution!

nomack84 commented:
Happy to help :)
